### PR TITLE
Fix superseding schema feature

### PR DESCRIPTION
**Pull request by stanch**
In the past, Iglu Server would take both `supersededBy` and `supersedes` as inputs, and would only output `supersededBy`. This caused two problems when schemas were promoted from DEV to PROD. Suppose 1-0-3 supersedes 1-0-2.
* If 1-0-2 is in PROD and 1-0-3 is in DEV, then when 1-0-3 is copied to PROD, there will be nothing to tell the PROD Iglu Server that it supersedes 1-0-2.
* If neither are in PROD, then when 1-0-2 is copied to PROD, there will be an error because 1-0-3 is still unknown to PROD Iglu Server.

This commit changes the behavior so that:
* It’s not possible to patch superseding relationships. It’s hardly ever necessary and would create all kinds of edge cases.
* Iglu Server only takes `supersedes` as an input (that’s ok, because using `supersededBy` requires patching anyway)
* Iglu Server outputs both `supersededBy` and `supersedes` (the `supersedes` value is fixed and not updated over time, but it’s enough for the promotion logic to work correctly)

### CODE REVIEW

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/InMemory.scala [hunk 1]
```diff
@@ -102,24 +102,38 @@ case class InMemory[F[_]](ref: Ref[F, InMemory.State]) extends Storage[F] {
...
   )(implicit F: Bracket[F, Throwable]): F[Unit] =
     for {
       db <- ref.get
-      schemas = superseded.toList.foldRight(db.schemas) {
-        case (v, m) =>
-          val supersededSchemaKey = SchemaMap(SchemaKey(vendor, name, "jsonschema", v))
-          m.updated(
-            supersededSchemaKey,
-            m(supersededSchemaKey).copy(supersededBy = Some(supersededBy))
-          )
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/InMemory.scala:122]: I think this was not propagating transitive relationships, e.g. if 1-0-2 supersedes 1-0-1 and now we say 1-0-3 supersedes 1-0-2, I don’t think this would update 1-0-1.

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala [hunk 2]
```diff
@@ -179,34 +176,61 @@ object Postgres {
     fr"superseded_by = ${version.asString}"
 
   def supersededFr(vendor: String, name: String, superseded: SchemaVer.Full): Fragment =
-    fr"(" ++ nameFr(vendor, name) ++ fr"AND" ++ schemaVerFr(superseded) ++ fr")" ++
-      fr"OR ${supersededByFr(superseded)}"
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala:183]: This was a bug — the second OR clause should have included the vendor+name clause as well!

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala [hunk 3]
```diff
@@ -179,34 +176,61 @@ object Postgres {
...
   object Sql {
+    def baseSchemaQuery(filter: Fragment, limit: Fragment = Fragment.empty): Query0[Schema] =
+      sql"""
+        SELECT $schemaColumns, STRING_AGG(s_version, ', ') AS supersedes
+        FROM $SchemasTable
+        LEFT OUTER JOIN (
+          SELECT vendor as s_vendor, name as s_name,
+            $versionToStringFr as s_version, superseded_by as s_superseded_by
+          FROM $SchemasTable
+        ) superseded
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala:202]: Open to suggestions on making this nicer... We could probably add some indexes and so on to speed up the query, but not sure if this is a real problem given a typical number of schemas.

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala [hunk 4]
```diff
@@ -265,16 +289,34 @@ object Postgres {
...
     ): Update0 =
-      (fr"UPDATE" ++ SchemasTable ++ fr"SET" ++ supersededByFr(supersededBy) ++ fr"WHERE" ++ supersededListFr(
-        vendor,
-        name,
-        superseded
-      )).update
+      (fr"UPDATE" ++ SchemasTable ++
+        Fragments.set(
+          supersededByFr(supersedingInfo.actualSupersedingVersion),
+          fr"updated_at = current_timestamp"
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala:297]: This was missing

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala [hunk 5]
```diff
@@ -265,16 +289,34 @@ object Postgres {
...
 
     def addSupersededByColumn: Update0 =
       (fr"ALTER TABLE" ++ SchemasTable ++ fr"ADD COLUMN IF NOT EXISTS superseded_by VARCHAR(32)").update
+
+    val addPrimaryKey: Update0 =
+      sql"""
+        DO $$$$ BEGIN
+          IF NOT EXISTS (
+            SELECT constraint_name
+            FROM information_schema.table_constraints
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala:314]: I believe the db user will automatically have access to information for tables it has access to. However, we might want to validate that before fully rolling out...

File: src/test/scala/com/snowplowanalytics/iglu/server/ServerSpec.scala [hunk 6]
```diff
@@ -175,7 +175,7 @@ class ServerSpec extends Specification {
     )
     val (expectedSchema100, _) = testSchema(
       version = SchemaVer.Full(1, 0, 0),
-      supersedingInfo = SupersedingInfo.SupersededBy(SchemaVer.Full(1, 0, 3)).some
+      supersedingInfo = None
```
Comment (stanch) [inline src/test/scala/com/snowplowanalytics/iglu/server/ServerSpec.scala:178]: No longer superseded, because `supersededBy` was used to patch it (we don’t respect that field anymore)

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala (hunk 5 above)
Comment (stanch) [reply src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala:314]: Also, applying a unique constraint to existing data is a bit scary. But I believe this constraint **must** hold? Otherwise we have way bigger problems.

File: src/test/scala/com/snowplowanalytics/iglu/server/service/SchemaServiceSpec.scala [hunk 7]
```diff
@@ -542,7 +542,7 @@ class SchemaServiceSpec extends org.specs2.Specification {
     )
     val (expectedSchema100, _) = testSchema(
       version = SchemaVer.Full(1, 0, 0),
-      supersedingInfo = Schema.SupersedingInfo.SupersededBy(SchemaVer.Full(1, 0, 1)).some
+      supersedingInfo = None
```
Comment (stanch) [inline src/test/scala/com/snowplowanalytics/iglu/server/service/SchemaServiceSpec.scala:545]: No longer superseded, because `supersededBy` was used to patch it (we don’t respect that field anymore)

**Review by stanch — COMMENTED** [review]
Test passed (on my machine), but I am not super comfortable with the current test coverage of the feature. The tests are somewhat hard to follow, and we seem to be testing the same behavior but in different flavors in different places.

I can think of a few new scenarios:
* Validate that superseding info is not confused across different vendors/names (I believe the code had a bug regarding to that)
* Dedicated tests for chaining patterns shown in the docs

Unfortunately I am not sure I’ll find the time to add those... :)

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/InMemory.scala (hunk 1 above)
Comment (stanch) [reply src/main/scala/com/snowplowanalytics/iglu/server/storage/InMemory.scala:122]: Actually there is something that bothers me here. Suppose you say 1-0-3 supersedes 1-0-2.

Currently, `Storage` is _not_ responsible for propagating _up_ the chain, i.e. ensuring that 1-0-4 is updated if it supersedes 1-0-3. Instead, `SchemaService` fetches the schema that supersedes 1-0-3 (if any) and tells `Storage` to update that one.

I am not sure how the `F[_]` stuff works, but it seems like the fetching and the updating might not run in the same transaction, so we potentially have a race condition where the update will not be correct. Or is Iglu Server single-threaded?

Further, `Storage` _is_ responsible for propagating _down_ the chain, i.e. ensuring that 1-0-1 is updated if 1-0-2 supersedes 1-0-1. It seems weird that it goes one way but not the other.

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala [hunk 8]
```diff
@@ -181,35 +195,69 @@ object Postgres {
...
+        .map(version =>
+          Fragments.or(
+            schemaVerFr(version),
+            supersededByFr(version)
+          )
+        )
+        .intercalate(fr"OR")
+    )
+
+  def supersededByInferiorFr(supersedingVersion: SchemaVer.Full) =
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/Postgres.scala:211]: This was also a bug (or at least an unhandled edge case) — a schema should not supersede one that’s already been superseded by a higher version

File: src/main/scala/com/snowplowanalytics/iglu/server/storage/InMemory.scala [hunk 9]
```diff
@@ -63,7 +64,16 @@ case class InMemory[F[_]](ref: Ref[F, InMemory.State]) extends Storage[F] {
...
   def getSchemas(implicit F: Bracket[F, Throwable]): F[List[Schema]] =
-    ref.get.map(state => state.schemas.values.toList.sortBy(_.metadata.createdAt))
+    ref
+      .get
+      .map(state =>
+        state
+          .schemas
+          .values
+          .toList
+          .sortBy(schema => (schema.metadata.createdAt, schema.schemaMap.schemaKey.version))
```
Comment (stanch) [inline src/main/scala/com/snowplowanalytics/iglu/server/storage/InMemory.scala:74]: Some tests were flaky because when you add schemas too quickly, they get the same timestamp

**Review by spenes — APPROVED** [review]
Everything looks great. Nothing to add really. Top-notch work, well done!

**Comment by spenes** [comment]
Moved [here](https://github.com/snowplow/iglu-server/pull/145). Closing this one.
